### ISSUE TITLE
Improve Uproot CodeGen to speed up Uproot transformer

### ISSUE THREAD

**Issue by kyungeonchoi** [ssl-hep/ServiceX#124]
# Story
The speed of Uproot transformer is much slower than ROOT `TTree::Draw()` method. Tests are performed using local files to exclude possible network issues. Exact numbers are hard to extract due to the caching for both methods. The rough number is about 5 times slower in the Uproot transformer than ROOT `TTree::Draw()` method for the selection using 47 branches. 
An example of the current Uproot transformer is shown by @kyungeonchoi at the [ServiceX meeting](https://indico.cern.ch/event/901356/contributions/3814958/attachments/2015249/3368271/KyungEon_3Apr2020.pdf). Possible improvements will be investigated by @kyungeonchoi, @masonproffitt, @gordonwatts.

# Test Conditions
1. [Uproot transformer reads a root file then returns Awkward Table after selection](https://github.com/ssl-hep/ServiceX_xAOD_CPP_transformer/blob/927554e5269fd47604be2e4f362906a0f274e94e/transformer_uproot.py#L129)
2. ROOT reads a root file and saves a histogram after selection
2. Selection touches 47 branches

# Acceptance Criteria
1. Uproot transformer provides a better performance

**Comment by masonproffitt** [comment]
I copied the exact code that KyungEon provided for both the ROOT `TTree::Draw()` method and the func_adl_uproot-generated function. The precise files that I used are at https://github.com/masonproffitt/func_adl_uproot_benchmark. The two procedures aren't exactly the same because the ROOT code saves a histogram while the uproot code does not, but I adjusted the uproot code to save a file with the full results of the selection to approximate this. I tested these on a modest desktop PC (AMD Phenom II X4 940 CPU, 8GiB RAM) with a fairly old and standard 5400 RPM HDD (Samsung HD204UI) using a pre-downloaded copy of the 22.1 GiB ROOT file `user.kchoi:ttHML_80fb_364250_mc16a.root`. For each, I ran five sets of five consecutive trials, clearing the disk cache via `echo 3 >/proc/sys/vm/drop_caches` between each set. These are my resulting running times (all in seconds):

ROOT `TTree::Draw()` macro:

|   | Set 1 | Set 2 | Set 3 | Set 4 | Set 5 | Average | Standard deviation |
|---|-------|-------|-------|-------|-------|---------|--------------------|
| **Trial 1** | 29.094 | 29.044 | 29.161 | 28.919 | 29.136 | 29.071 | 0.096 |
| **Trial 2** |  1.236 |  1.211 |  1.200 |  1.213 |  1.199 |  1.212 | 0.015 |
| **Trial 3** |  1.116 |  1.125 |  1.159 |  1.143 |  1.142 |  1.137 | 0.017 |
| **Trial 4** |  1.190 |  1.193 |  1.158 |  1.165 |  1.156 |  1.172 | 0.018 |
| **Trial 5** |  1.177 |  1.192 |  1.178 |  1.150 |  1.168 |  1.173 | 0.015 |

func_adl_uproot-generated code:

|   | Set 1 | Set 2 | Set 3 | Set 4 | Set 5 | Average | Standard deviation |
|---|-------|-------|-------|-------|-------|---------|--------------------|
| **Trial 1** | 15.530 | 15.263 | 15.414 | 15.413 | 15.519 | 15.428 | 0.108 |
| **Trial 2** |  3.068 |  3.056 |  3.037 |  3.039 |  3.086 |  3.057 | 0.021 |
| **Trial 3** |  3.038 |  3.036 |  3.111 |  3.052 |  3.207 |  3.089 | 0.073 |
| **Trial 4** |  3.037 |  3.011 |  3.512 |  3.042 |  3.319 |  3.184 | 0.222 |
| **Trial 5** |  3.065 |  3.019 |  3.511 |  3.057 |  3.502 |  3.231 | 0.252 |

The first trial is always significantly longer since the file is not cached. All subsequent trials are faster and have similar times. Before disk caching, I get 29.07 ± 0.10 s for `TTree::Draw()` vs. 15.43 ± 0.11 s for the uproot code. After disk caching, I get 1.17 ± 0.03 s for `TTree::Draw()` vs. 3.14 ± 0.17 s for the uproot code. Based on separate tests that I've done, approximately 2 s of the 3 s for running the uproot code after caching is unavoidable overhead of loading the uproot module and simply reading the branches necessary for selection.

**Comment by gordonwatts** [comment]
@kyungeonchoi does this match what you and @ponyisi are seeing?

**Comment by kyungeonchoi** [comment]
> @kyungeonchoi does this match what you and @ponyisi are seeing?

Unfortunately I found a problem with the local disk drive attached to the machine that I've been mostly used for the benchmarks. The command I've been using and also Mason used for his tests requires a linux machine and superuser privilege, and my linux machine apparently has some problem with the disk. This affects only to my remark above "The rough number is about 5 times slower in the Uproot transformer than ROOT TTree::Draw() method for the selection using 47 branches". I am currently doing a series of benchmarks using my stable laptop. I will update as soon as I have the results. Sorry for the delay!

**Comment by gordonwatts** [comment]
Thanks for the update, @kyungeonchoi. Though I would have thought that it would affect both of those in the same way.

Have you made any progress? :-)

**Comment by kyungeonchoi** [comment]
Tests performed using the same scripts with Mason (from his github). Due to the problem with my linux machine described above, tests are performed on my laptop (Mac, i5 CPU, 8 GB RAM). The input file `ttHML_80fb_364250_mc16a.root` (23.71 GB) is read from the external HDD (5400 RPM). The version of ROOT is 6.20/02, uproot==3.11.3, and awkward==0.12.20. The command used by Mason to empty the disk cache is not available on Mac, so I used the different tool to empty cache: `vmtouch`. This tool is also used by Jim Pivarski for his [uproot performance tests](https://indico.cern.ch/event/686641/contributions/2894906/attachments/1606247/2548596/pivarski-uproot.pdf). These are my results:

ROOT `TTree::Draw()` macro:
|         | Set 1  | Set2  | Set 3 | Set 4  |  Avg | SD |
| --------| ------ | ----- | ------|------- |----|----|
| Trial 1 | 9.74|9.74|9.68|9.63|9.70 | 0.05 |
| Trial 2 | 0.77|0.76|0.78|0.77|0.77 | 0.01 |
| Trial 3 | 0.76|0.78|0.78|0.77|0.77 | 0.01 |
| Trial 4 | 0.78|0.78|0.77|0.76|0.77 | 0.01 |
| Trial 5 | 0.78|0.77|0.77|0.77|0.77 | 0.01 |

func_adl_uproot-generated code:
|         | Set 1  | Set2  | Set 3 | Set 4  | Avg | SD |
| --------| ------ | ----- | ------|------- | ----|----|
| Trial 1 | 18.16|18.53|18.57|18.09|18.34 | 0.25 |
| Trial 2 | 1.93|1.93|1.90|1.89|1.91 | 0.02 |
| Trial 3 | 1.90|1.90|1.91|1.91|1.91 | 0.01 |
| Trial 4 |1.90|1.92|1.89|1.93|1.91 | 0.02 |
| Trial 5 | 1.89|1.93|1.88|1.91|1.90 | 0.02 |

Comparison with Mason's results:
- The first trials show about 3 times faster for the ROOT `TTree::Draw()` method and similar times for the func_adl_uproot-generated code. It's not clear why ROOT `TTree::Draw()` runs much faster than the func_adl_uproot-generated code for me, but slower for Mason. We might need to do a test using the same machine, e.g. lxplus. It should be something to do with reading or file cache.
- My results are slightly faster than Mason's results from the second trials (which are cached). It is due to the difference in the CPU performance (CPU benchmark score of my Macbook is quite higher than Mason's machine). The figure 1 shows the disk read of the external disk, and confirms the file is not read from the disk once it is cached.

Figure 1. Disk read after cached
![disk_io_cached](https://user-images.githubusercontent.com/54450665/80656181-2cab7780-8a46-11ea-89bc-d21dc067ddfe.png)


The following tests are done using the same machine but the file is read from the local NVMe SSD.

ROOT `TTree::Draw()` macro:
|         | Set 1  | Set2  | Set 3 | Set 4  |  Avg | SD |
| --------| ------ | ----- | ------|------- |----|----|
| Trial 1 | 1.68|1.72|1.7|1.59|1.67 | 0.06 |
| Trial 2 | 0.78|0.77|0.78|0.82|0.79 | 0.02 |
| Trial 3 | 0.77|0.79|0.78|0.79|0.78 | 0.01 |
| Trial 4 | 0.78|0.79|0.76|0.78|0.78 | 0.01 |
| Trial 5 | 0.77|0.78|0.78|0.77|0.78 | 0.01 |

func_adl_uproot-generated code:
|         | Set 1  | Set2  | Set 3 | Set 4  | Avg | SD |
| --------| ------ | ----- | ------|------- | ----|----|
| Trial 1 | 3.21|3.06|3.23|3.21| 3.18 | 0.08 |
| Trial 2 | 1.93|1.87|1.93|1.93|1.91 | 0.03 |
| Trial 3 | 1.93|1.9|1.92|1.89|1.91 | 0.02 |
| Trial 4 |1.92|1.91|1.91|1.9|1.91 | 0.01 |
| Trial 5 | 1.9|1.91|1.88|1.93|1.90 | 0.02 |

- The first trials (before cached) are significantly faster than the tests with the external HDD.
- After cached (trial 2-5), the times are basically the same with the external HDD since the file is already cached and no disk read required.

**Comment by gordonwatts** [comment]
This is great! Now, to recap, please tell me if I understand this right:

1. There are differences in uproot and TTDraw, with TTDraw being ~x2 faster when the cache is hot, about the same amount of time when the cache is cold.
1. The differences in cold cache could have to do with architecture, and are likely not worth following up for now.
1. The speed differences in the hot cache are probably interesting to follow up.

A question to @masonproffitt - you talk about the overhead of loading uproot and reading the branches. Doesn't the TTDraw test have to read the branches as well. Above you mention reading the branches - does that mean branch meta-data, perhaps? Which leads to a second question - do you test the time it takes from running the executable from scratch, or only once you are inside python? And the same question with root.

Am I correct that the initial reason for this issue is now done? We can close this? Or is there a large enough performance issue being seen we need to continue to track this down?

**Comment by kyungeonchoi** [comment]
Hi @gordonwatts , 
First of all, I would say that the initial reason of this study is not satisfied yet. The reason I started to investigate this is the Uproot transformation time at UC River increases faster than linear (blue line) as a function of input file size for very large (>20 GB) files as shown below. 
![VV_filesize_linearfit](https://user-images.githubusercontent.com/54450665/80678491-99dbfe80-8a80-11ea-9c12-ca3bc04415c7.png)
For example, a transformation of 72GB file is expected to be ~20 sec if it follows linear line as most of smaller files, but it took >160 sec.
In addition, the same process in the TRExFitter (essentially ROOT TTree::Draw()), i.e., read a flat ntuple and make a selection and return selected branch, runs noticeably faster than Uproot transformation at the lxplus machine. I think it is not fair to compare times between ROOT TTree::Draw() at lxplus and Uproot transformer at River for many reasons, but we can think of it as a reference.
It is clear that the goal we want to achieve is to improve transformation time especially for large files, and I've been investigating a possible room for improvement.

Regarding your number bullets,
1. It is not true that about the same amount of time when the cache is cold. Regardless of caching, it always takes about 2 times longer in my results for both HDD and SSD: 18.34/9.70 (cold), 1.91/0.77 (hot)
2. I believe we are not so interested in the hot cache since we don't quite want to store all inputs into the K8s node's cache (at least my use case doesn't need multiple access of the same input). I guess I haven't seen any sizable difference in time between consecutive transformations at UC River.

**Comment by gordonwatts** [comment]
Awesome!

Your comments about hot vs cold cache are good ones, I didn't think that through. Most of the time we will be running with a cold cache. As far as x2, yes, there is a x2, but the time is so small as I do not think it matters - 2 seconds vs 1 second is not... that big a deal. If it is 5 minutes vs 2.5 minutes, then, yes, it definitely would be worth pushing on.

That plot is great. This means now we need to figure out how to test in-situ up on River. I suppose the very first thing to do is try running on a pod, up on river, mostly outside the servicex environment. Just to see if a raw pod, where you run these test scripts, works or not.

@BenGalewsky - is this just a matter of starting a pod and copying everything up there? You probably know some cool kubernetes commands that would allow @kyungeonchoi to do that pretty easily... :-)

**Comment by masonproffitt** [comment]
> A question to @masonproffitt - you talk about the overhead of loading uproot and reading the branches. Doesn't the TTDraw test have to read the branches as well. Above you mention reading the branches - does that mean branch meta-data, perhaps? Which leads to a second question - do you test the time it takes from running the executable from scratch, or only once you are inside python? And the same question with root.

Yes, both methods have to read the branches involved in selection, of course. By "reading the branches", I really mean reading the whole branch (meta-data and all entries). In the repository that I linked, you can see that I test running both methods from scratch.

**Comment by masonproffitt** [comment]
Are the filesizes quoted in the above plot for the entire file and not just the TTree that the selection is being run on? If so, I think this is not the right metric to look at vs execution time. It's possible that there is a long seek time within the file for the very large filesizes, but that would be a very separate issue from anything directly related to func_adl_uproot.

I did some local tests on how the uproot code performs on varying TTree lengths. This is using the same `user.kchoi:ttHML_80fb_364250_mc16a.root` file from before. In order to perform a fair comparison between different lengths, I created a new file for each length N with only the `nominal` tree and with only the first N events of the original tree. This leads to filesizes ranging from 0.19 MiB (1 TTree entry) to 352 MiB (all 250758 TTree entries). Note that even the full TTree is quite small compared to the original file (22.1 GiB) with all the 78 other TTrees. These are my results:

![image](https://user-images.githubusercontent.com/32773304/80709005-e0ffd900-8aec-11ea-8aa2-3e7cdca1be98.png)

The line on here is a simple linear fit, which seems quite compatible.

I am more convinced than ever that func_adl_uproot is not where the problem lies. If there is an issue somewhere, it is probably either in seeking through these huge ROOT files or in the serialization step to Parquet.

**Comment by gordonwatts** [comment]
Very nice, thanks for this!

So all the more important to re-run this test up on `river` to eliminate that part of the system.

**Comment by gordonwatts** [comment]
What is the state of this? Are we finished here?

**Comment by kyungeonchoi** [comment]
@gordonwatts I wrote a simple PyROOT transformer and it can be modified to perform the same operation with uproot transformer. I'd like to do this comparison but no chance yet.

**Comment by gordonwatts** [comment]
Are we done with this issue, btw? I guess this is addressed to you, @kyungeonchoi. :-)

**Comment by BenGalewsky** [comment]
Assume closed
